Adds CoreDNS forward health check alerts

danehans opened the pull request:
Adds CoreDNS forward plugin health check alerts.

In support of [NE-358](https://issues.redhat.com/browse/NE-358)

danehans commented:
@povilasv can you PTAL at this PR when you have a moment?

lilic commented on alerts/forward.libsonnet:
```diff
@@ -51,6 +51,32 @@
+            expr: |||
+              rate(coredns_forward_healthcheck_failure_count_total[10s]) > 3
+            ||| % $._config,
+            'for': '30s',
```
I would rather make this be 10 minutes. Any reason why 30 seconds?

danehans replied:
@lilic my thinking is that if this rate is achieved, it can be a pretty significant issue for apps that require the upstream for name resolution. However, I'll change to 10m based on your guidance.

danehans replied:
Commit `5cbc692` addresses your comment here.

lilic commented on alerts/forward.libsonnet:
```diff
@@ -51,6 +51,32 @@
+            expr: |||
+              rate(coredns_forward_healthcheck_broken_count_total[5s]) > 3
+            ||| % $._config,
+            'for': '15s',
```
Same as above, 10 minutes or 15 minutes for cause is more common.

danehans replied:
Same as https://github.com/povilasv/coredns-mixin/pull/6#discussion_r470114108

danehans replied:
Commit `5cbc692` addresses your comment here.

lilic commented on alerts/forward.libsonnet:
```diff
@@ -51,6 +51,32 @@
+          {
+            alert: 'CoreDNSForwardHealthcheckFailureCount',
+            expr: |||
+              rate(coredns_forward_healthcheck_failure_count_total[10s]) > 3
```
We should add a `%(corednsSelector)s` selector here, like all the above alerts have. Also lets make this be 5 minute range selector instead of 10 seconds.

danehans replied:
Commit `5cbc692` addresses your comment here.

danehans commented on alerts/forward.libsonnet:
```diff
@@ -51,6 +51,32 @@
+          {
+            alert: 'CoreDNSForwardHealthcheckFailureCount',
+            expr: |||
+              rate(coredns_forward_healthcheck_failure_count_total[10s]) > 3
```
@paulfantom thanks for the review and link to help me better understand rate. Commit `5cbc692` addresses your comment here.

danehans commented on alerts/forward.libsonnet:
```diff
@@ -51,6 +51,32 @@
+          {
+            alert: 'CoreDNSForwardHealthcheckBrokenCount',
+            expr: |||
+              rate(coredns_forward_healthcheck_broken_count_total[5s]) > 3
```
Commit `5cbc692` addresses your comment here.

danehans commented:
@povilasv I added tests but they're failing:
```
promtool test rules tests.yaml
Unit Testing:  tests.yaml
  FAILED:
    alertname:CoreDNSForwardHealthcheckFailureCount, time:11m, 
        exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckFailureCount\", severity=\"warning\"} Annotations:{message=\"100 CoreDNS health checks have failed to upstream server 5.6.7.8 on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckfailurecount\"}]", 
        got:"[]"
    alertname:CoreDNSForwardHealthcheckBrokenCount, time:11m, 
        exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckBrokenCount\", severity=\"warning\"} Annotations:{message=\"100 CoreDNS health checks have failed for all upstream servers on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckbrokencount\"}]", 
        got:"[]"

make: *** [test] Error 1
```

Does the above failure indicate that the alert rule did not fire?

povilasv commented on tests.yaml:
```diff
@@ -44,3 +44,31 @@ tests:
+- interval: 1m
+  input_series:
+    - series: 'coredns_forward_healthcheck_failure_count_total{instance="1.2.3.4:9153",job="kube-dns",kubernetes_name="coredns-kube-metrics",kubernetes_namespace="kube-system",pod="coredns-65b6759cb4-qgdxp",server="dns://:53",zone=".",to="5.6.7.8"}'
+      values: '100'
```
add more values as right now you have only a value at t=0, t=1m will have no value.

povilasv commented on tests.yaml:
```diff
@@ -44,3 +44,31 @@ tests:
+    - series: 'coredns_forward_healthcheck_broken_count_total{instance="1.2.3.4:9153",job="kube-dns",kubernetes_name="coredns-kube-metrics",kubernetes_namespace="kube-system",pod="coredns-65b6759cb4-qgdxp",server="dns://:53",zone=".",to="5.6.7.8"}'
+      values: '100'
+  alert_rule_test:
+    - eval_time: 11m
```
ditto

povilasv commented:
> @povilasv I added tests but they're failing:
> 
> ```
> promtool test rules tests.yaml
> Unit Testing:  tests.yaml
>   FAILED:
>     alertname:CoreDNSForwardHealthcheckFailureCount, time:11m, 
>         exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckFailureCount\", severity=\"warning\"} Annotations:{message=\"100 CoreDNS health checks have failed to upstream server 5.6.7.8 on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckfailurecount\"}]", 
>         got:"[]"
>     alertname:CoreDNSForwardHealthcheckBrokenCount, time:11m, 
>         exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckBrokenCount\", severity=\"warning\"} Annotations:{message=\"100 CoreDNS health checks have failed for all upstream servers on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckbrokencount\"}]", 
>         got:"[]"
> 
> make: *** [test] Error 1
> ```
> 
> Does the above failure indicate that the alert rule did not fire?

Yes I've added the comment why ;)

danehans commented:
@povilasv I updated the values for both alerts but I'm getting the same test failure as https://github.com/povilasv/coredns-mixin/pull/6#issuecomment-676659245. I also updated the metric names since I found the following information in the coredns 1.7.0 release notes:
```
### Metric Changes

We mostly dropped `count` from `_total` metrics names:
<SNIP>
* `coredns_forward_healthcheck_failure_count_total` -\> `coredns_forward_healthcheck_failures_total`
* `coredns_forward_healthcheck_broken_count_total` -\> `coredns_forward_healthcheck_broken_total`
```

Since [PR 4025](https://github.com/coredns/coredns/pull/4025), CoreDNS uses the [promauto](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/promauto?tab=doc) package to ensure all created metrics are properly registered. Could this be related to the test failures?

povilasv commented:
> @povilasv I updated the values for both alerts but I'm getting the same test failure as [#6 (comment)](https://github.com/povilasv/coredns-mixin/pull/6#issuecomment-676659245). I also updated the metric names since I found the following information in the coredns 1.7.0 release notes:
> 
> ```
> ### Metric Changes
> 
> We mostly dropped `count` from `_total` metrics names:
> <SNIP>
> * `coredns_forward_healthcheck_failure_count_total` -\> `coredns_forward_healthcheck_failures_total`
> * `coredns_forward_healthcheck_broken_count_total` -\> `coredns_forward_healthcheck_broken_total`
> ```
> 
> Since [PR 4025](https://github.com/coredns/coredns/pull/4025), CoreDNS uses the [promauto](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/promauto?tab=doc) package to ensure all created metrics are properly registered. Could this be related to the test failures?

Unrelated. Tests look at only fake data you define in https://github.com/povilasv/coredns-mixin/pull/6/files#diff-d5d183d43dbaa81a9750464ef604af3eR51 and execute PromQL on top of fake data.

Now your alerts trigger on ` sum(rate(coredns_forward_healthcheck_failures_total{%(corednsSelector)s}[5m])) > 3` for 10 mins.

Which means: failures per second needs to be higher than 3 looking at last window of 5 mins for 10 mins.

your fake data: `0 1 2 3 4 5 6 7 8 9 10 11 12`, interval 1 min means it has a rate of 1 per minute, which is 1/60 per second, which will never fire.

Try changing your alert to ` sum(rate(coredns_forward_healthcheck_failures_total{%(corednsSelector)s}[5m])) > 0`

danehans commented:
@povilasv thanks for the additional background. I think I'm starting to connect the dots. I made the change as you suggested, but still no luck:
```
$ git diff
diff --git a/alerts/forward.libsonnet b/alerts/forward.libsonnet
index 7485de5..af1acab 100644
--- a/alerts/forward.libsonnet
+++ b/alerts/forward.libsonnet
@@ -54,7 +54,7 @@
           {
             alert: 'CoreDNSForwardHealthcheckFailureCount',
             expr: |||
-              sum(rate(coredns_forward_healthcheck_failures_total{%(corednsSelector)s}[5m])) > 3
+              sum(rate(coredns_forward_healthcheck_failures_total{%(corednsSelector)s}[5m])) > 0
             ||| % $._config,
             'for': '10m',
             labels: {
@@ -67,7 +67,7 @@
           {
             alert: 'CoreDNSForwardHealthcheckBrokenCount',
             expr: |||
-              sum(rate(coredns_forward_healthcheck_broken_total[5m])) > 3
+              sum(rate(coredns_forward_healthcheck_broken_total[5m])) > 0
             ||| % $._config,
             'for': '10m',
             labels: {

$ make all
find . -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | \
		xargs -n 1 -- jsonnetfmt  -i
find . -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | \
		while read f; do \
			jsonnetfmt  "$f" | diff -u "$f" -; \
		done
promtool check rules prometheus_alerts.yaml
Checking prometheus_alerts.yaml
  SUCCESS: 9 rules found

promtool test rules tests.yaml
Unit Testing:  tests.yaml
  FAILED:
    alertname:CoreDNSForwardHealthcheckFailureCount, time:11m, 
        exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckFailureCount\", severity=\"warning\"} Annotations:{message=\"CoreDNS health checks have failed to upstream server 5.6.7.8 on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckfailurecount\"}]", 
        got:"[]"
    alertname:CoreDNSForwardHealthcheckBrokenCount, time:11m, 
        exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckBrokenCount\", severity=\"warning\"} Annotations:{message=\"CoreDNS health checks have failed for all upstream servers on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckbrokencount\"}]", 
        got:"[]"

make: *** [test] Error 1
```

povilasv commented on tests.yaml:
```diff
@@ -44,3 +44,31 @@ tests:
+
+- interval: 1m
+  input_series:
+  - series: 'coredns_forward_healthcheck_failure_count_total{instance="1.2.3.4:9153",job="kube-dns",kubernetes_name="coredns-kube-metrics",kubernetes_namespace="kube-system",pod="coredns-65b6759cb4-qgdxp",server="dns://:53",zone=".",to="5.6.7.8"}'
```
```suggestion
  - series: 'coredns_forward_healthcheck_failures_total{instance="1.2.3.4:9153",job="kube-dns",kubernetes_name="coredns-kube-metrics",kubernetes_namespace="kube-system",pod="coredns-65b6759cb4-qgdxp",server="dns://:53",zone=".",to="5.6.7.8"}'
```

povilasv commented on tests.yaml:
```diff
@@ -44,3 +44,31 @@ tests:
+
+- interval: 1m
+  input_series:
+  - series: 'coredns_forward_healthcheck_broken_count_total{instance="1.2.3.4:9153",job="kube-dns",kubernetes_name="coredns-kube-metrics",kubernetes_namespace="kube-system",pod="coredns-65b6759cb4-qgdxp",server="dns://:53",zone=".",to="5.6.7.8"}'
```
```suggestion
  - series: 'coredns_forward_healthcheck_broken_total{instance="1.2.3.4:9153",job="kube-dns",kubernetes_name="coredns-kube-metrics",kubernetes_namespace="kube-system",pod="coredns-65b6759cb4-qgdxp",server="dns://:53",zone=".",to="5.6.7.8"}'
```

povilasv commented:
> @povilasv thanks for the additional background. I think I'm starting to connect the dots. I made the change as you suggested, but still no luck:
> 
> ```
> $ git diff
> diff --git a/alerts/forward.libsonnet b/alerts/forward.libsonnet
> index 7485de5..af1acab 100644
> --- a/alerts/forward.libsonnet
> +++ b/alerts/forward.libsonnet
> @@ -54,7 +54,7 @@
>            {
>              alert: 'CoreDNSForwardHealthcheckFailureCount',
>              expr: |||
> -              sum(rate(coredns_forward_healthcheck_failures_total{%(corednsSelector)s}[5m])) > 3
> +              sum(rate(coredns_forward_healthcheck_failures_total{%(corednsSelector)s}[5m])) > 0
>              ||| % $._config,
>              'for': '10m',
>              labels: {
> @@ -67,7 +67,7 @@
>            {
>              alert: 'CoreDNSForwardHealthcheckBrokenCount',
>              expr: |||
> -              sum(rate(coredns_forward_healthcheck_broken_total[5m])) > 3
> +              sum(rate(coredns_forward_healthcheck_broken_total[5m])) > 0
>              ||| % $._config,
>              'for': '10m',
>              labels: {
> 
> $ make all
> find . -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | \
> 		xargs -n 1 -- jsonnetfmt  -i
> find . -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | \
> 		while read f; do \
> 			jsonnetfmt  "$f" | diff -u "$f" -; \
> 		done
> promtool check rules prometheus_alerts.yaml
> Checking prometheus_alerts.yaml
>   SUCCESS: 9 rules found
> 
> promtool test rules tests.yaml
> Unit Testing:  tests.yaml
>   FAILED:
>     alertname:CoreDNSForwardHealthcheckFailureCount, time:11m, 
>         exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckFailureCount\", severity=\"warning\"} Annotations:{message=\"CoreDNS health checks have failed to upstream server 5.6.7.8 on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckfailurecount\"}]", 
>         got:"[]"
>     alertname:CoreDNSForwardHealthcheckBrokenCount, time:11m, 
>         exp:"[Labels:{alertname=\"CoreDNSForwardHealthcheckBrokenCount\", severity=\"warning\"} Annotations:{message=\"CoreDNS health checks have failed for all upstream servers on 1.2.3.4:9153.\", runbook_url=\"https://github.com/povilasv/coredns-mixin/tree/master/runbook.md#alert-name-corednsforwardhealthcheckbrokencount\"}]", 
>         got:"[]"
> 
> make: *** [test] Error 1
> ```

you have a wrong metric in the tests :)

danehans commented:
@povilasv thanks for all your support with this PR. I've been tied-up with other priorities, but I pushed commit `5531a15` to implement your suggestions. PTAL when you have a moment.

danehans commented:
@povilasv following-up on this PR. Can you PTAL when you have a moment?

danehans commented:
Closing since the repo is not being maintained.
